q2cli/tests/test_tools.py

```diff
@@ -15,10 +15,97 @@
 from qiime2 import Artifact
 from qiime2.core.testing.util import get_dummy_plugin
 
-from q2cli.builtin.tools import tools
+from q2cli.builtin.tools import tools, _load_metadata
 from q2cli.commands import RootCommand
 
 
+class TestCastMetadata(unittest.TestCase):
+    def setUp(self):
+        self.runner = CliRunner()
+        self.tempdir = tempfile.mkdtemp(prefix='qiime2-q2cli-test-temp-')
+
+        self.metadata_file = os.path.join(
+                self.tempdir, 'metadata.tsv')
+        with open(self.metadata_file, 'w') as f:
+            f.write('id\tnumbers\tstrings\n0\t42\tabc\n1\t-1.5\tdef\n')
+
+        self.output_file = os.path.join(
+                self.tempdir, 'test_output.tsv')
+
+    def test_input_invalid_column_type(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:foo', '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn('Unknown column type provided.', result.output)
+
+    def test_input_duplicate_columns(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:numerical', '--cast', 'numbers:categorical',
+                    '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn('appears in cast more than once.', result.output)
```

Comment:
Fabu! Can you also check that the offending column name is correctly captured in the error message?